### PR TITLE
Added body/cell/column/column-type/column-list d.ts

### PR DESCRIPTION
**Pull request by munkk**

### CODE REVIEW

File: src/core/body/body.view.d.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {View} from '../view/view';
+
+
+export declare class BodyView extends View {
+	constructor(model: object, table: object);
+
+	table: object;
```
Comment (klumba12) [inline src/core/body/body.view.d.ts]: remove table and columnList, invalidate
add columns, rowspan, colspan, value, label

File: src/core/cell/cell.d.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+	set value(value: string);
+
+	get label(): object;
+
+	set label(label: object);
+
+	mode(value: string): void;
+
+	static equals(x: any, y: any): boolean;
```
Comment (klumba12) [inline src/core/cell/cell.d.ts]: remove equals

File: src/core/cell/cell.selector.d.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export declare class CellSelector {
+
+	constructor(model: object, table: object);
+
+	map<V>(items: any[]): (any) => V;
```
Comment (klumba12) [inline src/core/cell/cell.selector.d.ts]: returns cell type

File: src/core/cell/cell.selector.d.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export declare class CellSelector {
+
+	constructor(model: object, table: object);
```
Comment (klumba12) [inline src/core/cell/cell.selector.d.ts]: model: Model, table: Table

File: src/core/column-list/column.list.generate.d.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import {ColumnListModel} from "../column-list/column.list.model";
+import {ICreateResult} from "../column/column.factory";
+import {RowDetails} from "../row-details/row.details";
+
+export declare interface IColumnFactory{
+	(model: ColumnListModel): ICreateResult;
+}
+
+export declare function generate(rows: RowDetails[], columnFactory: IColumnFactory, deep: boolean): any;
```
Comment (klumba12) [inline src/core/column-list/column.list.generate.d.ts]: rows: any[],
I'm not sure about IColumnFActory
returns list of columns

File: src/core/column-list/column.list.model.d.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import {ColumnModel} from "../column-type/column.model";
+
+export declare class ColumnListModel {
+	constructor();
+
+	generation: object; // deep | shallow
```
Comment (klumba12) [inline src/core/column-list/column.list.model.d.ts]: generation is string

File: src/core/column-list/column.list.model.d.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+	index: any[];
```
Comment (klumba12) [inline src/core/column-list/column.list.model.d.ts]: array of strings

File: src/core/body/body.view.d.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import {View} from '../view/view';
+import {ColumnListModel} from "../column-list/column.list.model";
+import {ColumnModel} from "../column-type/column.model";
+
+export declare class BodyView extends View {
+	constructor(model: object, table: object);
```
Comment (klumba12) [inline src/core/body/body.view.d.ts]: model should be Model type and table Table type

File: src/core/body/body.view.d.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+	table: object;
+	rows: any[];
+
+	columns(row: any, pin: string): ColumnModel[];
+
+	rowspan(): number;
```
Comment (klumba12) [inline src/core/body/body.view.d.ts]: Please add(column, row) as arguments (for the future)

File: src/core/body/body.view.d.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+	colspan(column: ColumnModel, row: any): number;
+
+	value(row: any, column: ColumnModel, value: string): string;
```
Comment (klumba12) [inline src/core/body/body.view.d.ts]: value can be any

File: src/core/cell/cell.selector.d.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import {Model} from "../infrastructure/model";
+import {Table} from "../dom/table";
+import {Cell} from "./cell";
+export declare class CellSelector {
+
+	constructor(model: Model, table: Table);
+
+	map(items: any[]): Cell;
```
Comment (klumba12) [inline src/core/cell/cell.selector.d.ts]: return array of Cells

File: src/core/column-list/column.list.generate.d.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import {ColumnListModel} from "../column-list/column.list.model";
+import {ICreateResult} from "../column/column.factory";
+
+
+export declare interface IColumnFactory{
+	(model: ColumnListModel): ICreateResult;
```
Comment (klumba12) [inline src/core/column-list/column.list.generate.d.ts]: ICreateResult is a bad name too common

File: src/core/column-type/array.column.d.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+
+export declare class ArrayColumnModel extends DataColumnModel {
+	constructor();
+}
+
+export declare class ArrayColumn extends ColumnView {
+	constructor(model: ColumnModel);
+
+	static model(model: ColumnModel): any;
```
Comment (klumba12) [inline src/core/column-type/array.column.d.ts]: As columnView contains 'model' don't need to declare it in any column class

File: src/core/column-type/data.column.model.d.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+export declare interface IAggregationOptions{
+	distinct: boolean;
+	separator: string;
+}
+
+export class DataColumnModel extends ColumnModel {
+	constructor();
+
+	isDefault: boolean;
+	aggregation: object
```
Comment (klumba12) [inline src/core/column-type/data.column.model.d.ts]: aggregation is string

File: src/core/column-type/email.column.d.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import {ColumnView} from './column.model.view';
+import {DataColumnModel} from './data.column.model';
+import {ColumnModel} from "./column.model";
+
+export declare interface IEditorOptions {
+	trigger: string;
+	modelFactory: Function;
```
Comment (klumba12) [inline src/core/column-type/email.column.d.ts]: We decided to not use Function ? should be () => Model

File: src/core/column-type/password.column.d.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import {ColumnView} from './column.model.view';
+import {DataColumnModel} from './data.column.model';
+import {ColumnModel} from "./column.model";
+
+
+export declare class PasswordColumnModel extends DataColumnModel {
+	constructor();
+
+	canSort: boolean;
```
Comment (klumba12) [inline src/core/column-type/password.column.d.ts]: This is declared in base class don't do it here same for canFilter

File: src/core/column-type/pivot.column.d.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
+import {ColumnView} from './column.model.view';
+import {ColumnModel} from './column.model';
+
+
+export declare class PivotColumnModel extends ColumnModel {
+	constructor();
+
+	key: string;
```
Comment (klumba12) [inline src/core/column-type/pivot.column.d.ts]: defined in base class all only add rowIndex?

File: src/core/column/column.service.d.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+export declare function find(columns: ColumnModel[], key: string): ColumnModel;
+
+export declare function findIndex(columns: ColumnModel[], key: string): number;
+
+export declare function findView(columns: ColumnModel[], key: string): ColumnModel;
+
+export declare function dataView(columns: ColumnModel[], model: object): any[];
+
+export declare function lineView(columnRows: any[]): any[];
```
Comment (klumba12) [inline src/core/column/column.service.d.ts]: array of ColumnView?

File: src/core/column/column.service.d.ts [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+
+export declare function widthFactory(model: object, form: object): any;
```
Comment (klumba12) [inline src/core/column/column.service.d.ts]: model is Model

File: src/core/cell/cell.d.ts [hunk 20]
```diff
@@ -0,0 +1,22 @@
+export declare interface ICell{
+	entry: ICell;
```
Comment (klumba12) [inline src/core/cell/cell.d.ts]: Remove entry propertt
Y

File: src/core/cell/cell.d.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+	constructor(cell: ICell);
+
+	get value(): any ;
+
+	set value(value: any);
+
+	get label(): object;
+
+	set label(label: object);
+
```
Comment (klumba12) [inline src/core/cell/cell.d.ts]: Setter returns nothing)

File: src/core/cell/cell.d.ts [hunk 22]
```diff
@@ -0,0 +1,22 @@
+export declare interface ICell{
+	entry: ICell;
+	rowIndex: number;
+	columnIndex: number;
+	column: object;
```
Comment (titovmx) [inline src/core/cell/cell.d.ts]: ```column``` is a ColumnModel

File: src/core/cell/cell.d.ts [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+	row: object;
```
Comment (titovmx) [inline src/core/cell/cell.d.ts]: row should be ```any``` like other places

File: src/core/cell/cell.d.ts [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+export declare class Cell {
+
+	constructor(cell: ICell);
+
+	get value(): any ;
+
+	set value(value: any);
+
+	get label(): object;
```
Comment (titovmx) [inline src/core/cell/cell.d.ts]: ```label``` should be ```any``` too
If label isn't set by user, it returns value

File: src/core/cell/cell.d.ts [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+
+	set label(label: object);
+
+	mode(value: any): void;
```
Comment (titovmx) [inline src/core/cell/cell.d.ts]: should be ```mode(value: string)```

File: src/core/body/body.view.d.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+	table: object;
+	rows: any[];
+
+	columns(row: any, pin: string): ColumnModel[];
+
+	rowspan(): number;
+
+	colspan(column: ColumnModel, row: any): number;
+
+	value(row: any, column: ColumnModel, value: any): string;
```
Comment (vkorolev) [inline src/core/body/body.view.d.ts]: There are also `valueFactory` and `labelFactory` methods

File: src/core/column/column.service.d.ts [hunk 27]
```diff
@@ -0,0 +1,20 @@
+import {ColumnModel} from '../column-type/column.model';
+import {ColumnView} from "./column.view";
+import {Model} from "../infrastructure/model";
+
+
+export declare function map(columns: ColumnModel[]): any;
```
Comment (vkorolev) [inline src/core/column/column.service.d.ts]: It's not any. Something like `[key: string]: ColumnModel`

File: src/core/column/column.service.d.ts [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+
+export declare function getValue(column: ColumnModel): string;
+
+export declare function find(columns: ColumnModel[], key: string): ColumnModel;
```
Comment (vkorolev) [inline src/core/column/column.service.d.ts]: Return value may also be null

File: src/core/column-type/column.model.d.ts [hunk 29]
```diff
@@ -0,0 +1,32 @@
+import {IEditorOptions} from "./email.column";
```
Comment (titovmx) [inline src/core/column-type/column.model.d.ts]: please align declarations, now indents are different - somewhere spaces, somewhere tabs. In the project we uses tabs by convention now

File: src/core/column/column.service.d.ts [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+
+export declare function findIndex(columns: ColumnModel[], key: string): number;
+
+export declare function findView(columns: ColumnModel[], key: string): ColumnModel;
+
+export declare function dataView(columns: ColumnModel[], model: Model): any[];
```
Comment (vkorolev) [inline src/core/column/column.service.d.ts]: Array of `ColumnModel`

File: src/core/column/column.service.d.ts [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+
+export declare function lineView(columnRows: ColumnView[]): any[];
+
+export declare function widthFactory(model: Model, form: object): any;
```
Comment (vkorolev) [inline src/core/column/column.service.d.ts]: Not any. It's something like `c: ColumnModel => number`

File: src/core/body/body.view.d.ts [hunk 32]
```diff
@@ -0,0 +1,21 @@
+import {View} from '../view/view';
```
Comment (vkorolev) [inline src/core/body/body.view.d.ts:1]: There's no type definition for `../view/view`. Please add one

File: src/core/column-type/email.column.d.ts [hunk 33]
```diff
@@ -0,0 +1,18 @@
+import {ColumnView} from './column.model.view';
+import {DataColumnModel} from './data.column.model';
+import {Model} from "../infrastructure/model";
+
+export declare interface IEditorOptions {
```
Comment (titovmx) [inline src/core/column-type/email.column.d.ts]: ```IEditorOptions``` should be in separate file or in column.model at least

File: src/core/column-type/column.model.d.ts [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+export declare class ColumnModel {
+
+	constructor(type: string);
+
+	type: string;
+	key: string;
+	title: string;
+	value: string;
```
Comment (titovmx) [inline src/core/column-type/column.model.d.ts]: ```value``` is a function here
it's a custom getter which can be set by user

File: src/core/column-type/column.model.d.ts [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+   canEdit: boolean;
+	canResize: boolean;
+	canSort: boolean;
+	canMove: boolean;
+	canFilter: boolean;
+	canHighlight: boolean;
+	canFocus: boolean;
+   isVisible: boolean;
+	index: boolean;
+   label: string;
```
Comment (titovmx) [inline src/core/column-type/column.model.d.ts]: ```label``` is a function too

File: src/core/column-type/column.model.d.ts [hunk 36]
```diff
@@ -0,0 +1,32 @@
...
+	maxWidth: number;
+   canEdit: boolean;
+	canResize: boolean;
+	canSort: boolean;
+	canMove: boolean;
+	canFilter: boolean;
+	canHighlight: boolean;
+	canFocus: boolean;
+   isVisible: boolean;
+	index: boolean;
```
Comment (titovmx) [inline src/core/column-type/column.model.d.ts]: ```index``` is a ```number```

File: src/core/column-type/email.column.d.ts [hunk 37]
```diff
@@ -0,0 +1,18 @@
...
+	trigger?: string;
+	modelFactory?: () => Model;
+	label?: string;
```
Comment (titovmx) [inline src/core/column-type/email.column.d.ts]: this ```label``` is a function too

File: src/core/column-list/column.list.model.d.ts [hunk 38]
```diff
@@ -0,0 +1,11 @@
+import {ColumnModel} from "../column-type/column.model";
+
+export declare class ColumnListModel {
+	constructor();
+
+	generation: string; // deep | shallow
```
Comment (vkorolev) [inline src/core/column-list/column.list.model.d.ts]: `generation` may be enum

File: src/core/column-type/file.column.d.ts [hunk 39]
```diff
@@ -0,0 +1,20 @@
...
+
+export declare interface IHasPreview {
+	(name: string): boolean;
+}
+
+export declare class FileColumnModel extends DataColumnModel {
+	constructor();
+
+	canUpload?: boolean;
+	editorOptions?: IEditorOptions;
```
Comment (titovmx) [inline src/core/column-type/file.column.d.ts]: ```editorOptions``` had already declared in ColumnModel

File: src/core/column-type/file.column.d.ts [hunk 40]
```diff
@@ -0,0 +1,20 @@
+import {ColumnView} from './column.model.view';
+import {DataColumnModel} from './data.column.model';
+import {IEditorOptions} from './email.column';
+
+export declare interface IHasPreview {
```
Comment (titovmx) [inline src/core/column-type/file.column.d.ts]: I think separate file for ```IHasPreview``` will be better too

Comment (klumba12) [reply src/core/column-type/file.column.d.ts]: In this case it's only return function interface, it'll be overkill to have separate files for that

File: src/core/body/body.view.d.ts (hunk 32 above)
Comment (munkk) [reply src/core/body/body.view.d.ts:1]: there are no type definition for most of the files, they will be added as soon as possible)

File: src/core/column/column.service.d.ts [hunk 41]
```diff
@@ -0,0 +1,26 @@
...
+
+export declare function find(columns: ColumnModel[], key: string): ReturnColumnModelOrNull;
+
+export declare function findIndex(columns: ColumnModel[], key: string): number;
+
+export declare function findView(columns: ColumnModel[], key: string): ColumnModel[];
+
+export declare function dataView(columns: ColumnModel[], model: Model): ColumnModel[];
+
+export declare function lineView(columnRows: ColumnView[]): any[];
```
Comment (klumba12) [inline src/core/column/column.service.d.ts]: returns not any

File: src/core/column/column.service.d.ts [hunk 42]
```diff
@@ -0,0 +1,26 @@
...
+
+export declare function widthFactory(model: Model, form: object): any;
```
Comment (klumba12) [inline src/core/column/column.service.d.ts]: returns not any

File: src/core/column/column.service.d.ts [hunk 43]
```diff
@@ -0,0 +1,26 @@
...
+export declare type ReturnStringOrNull = string | null;
+export declare type ReturnColumnModelOrNull = ColumnModel | null;
+
+export declare class IMapResult{
+	[key: string]: ColumnModel;
+}
+
+export declare function map(columns: ColumnModel[]): IMapResult;
+
+export declare function getValue(column: ColumnModel): ReturnStringOrNull;
```
Comment (klumba12) [inline src/core/column/column.service.d.ts:14]: I'm not sure about Return values

File: src/core/column/column.factory.d.ts [hunk 44]
```diff
@@ -0,0 +1,7 @@
+import {Model} from "../infrastructure/model";
+
+export declare interface IEntity{
+	(entityType: string, columnType: string, body: string ): any;
```
Comment (klumba12) [inline src/core/column/column.factory.d.ts]: not any

File: src/core/column-type/url.column.d.ts [hunk 45]
```diff
@@ -0,0 +1,15 @@
+import {ColumnView} from './column.model.view';
+import {DataColumnModel} from './data.column.model';
+import {IEditorOptions} from './column.model';
+import {ColumnModel} from "./column.model";
+
+
+export declare class UrlColumnModel extends DataColumnModel {
+	constructor();
+
+   editorOptions?: IEditorOptions;
```
Comment (klumba12) [inline src/core/column-type/url.column.d.ts]: it should be in base class?

File: src/core/column-type/column.model.view.d.ts [hunk 46]
```diff
@@ -0,0 +1,13 @@
+import {View} from '../view/view';
+import {Model} from "../infrastructure/model";
+
+export declare class ColumnView extends View {
+	constructor(model: Model);
+
+	colspan?: number;
```
Comment (klumba12) [inline src/core/column-type/column.model.view.d.ts]: why '?'

File: src/core/body/body.view.d.ts [hunk 47]
```diff
@@ -0,0 +1,31 @@
...
+	constructor(public model: Model, public table: Table);
+
+	table: Table;
+	rows: RowModel[];
+
+	columns(row: any, pin: string): ColumnModel[];
+
+	rowspan(row: any, column: ColumnModel): number;
+
+	colspan(column: ColumnModel, row: any): number;
```
Comment (klumba12) [inline src/core/body/body.view.d.ts:22]: Dont row goes first?

**Review by klumba12 — APPROVED** [review]

**Pull request merged**